diff --git a/build.zig b/build.zig
index 67250bc..acf4ba1 100644
--- a/build.zig
+++ b/build.zig
@@ -50,9 +50,14 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
     lib.root_module.addCMacro("OPENSSL_NO_KTLS", "");
     lib.root_module.addCMacro("OPENSSL_NO_QUIC", "");
     lib.root_module.addCMacro("OPENSSL_NO_THREAD_POOL", "");
-    lib.root_module.addCMacro("OPENSSL_NO_STDIO", "");
     lib.root_module.addCMacro("OPENSSL_NO_JITTER", "");
+    lib.root_module.addCMacro("OPENSSL_NO_SM4", "");
+    lib.root_module.addCMacro("OPENSSL_NO_ML_DSA", "");
+    lib.root_module.addCMacro("OPENSSL_NO_SLH_DSA", "");
+    lib.root_module.addCMacro("OPENSSL_NO_ML_KEM", "");
     lib.root_module.addCMacro("OPENSSLDIR", "\"/usr/local/ssl\"");
+    lib.root_module.addCMacro("ENGINESDIR", "\"/usr/local/lib/engines\"");
+    lib.root_module.addCMacro("MODULESDIR", "\"/usr/local/lib/ossl-modules\"");
     lib.root_module.addCMacro("OSSL_PKEY_PARAM_RSA_DERIVE_FROM_PQ", "1");
     if (lib.rootModuleTarget().isMinGW())
         lib.root_module.addCMacro("NOCRYPT", "1");
@@ -312,6 +317,7 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/comp/c_zstd.c",
             "crypto/comp/comp_err.c",
             "crypto/comp/comp_lib.c",
+            "crypto/comp_methods.c",
             "crypto/conf/conf_api.c",
             "crypto/conf/conf_def.c",
             "crypto/conf/conf_err.c",
@@ -342,6 +348,7 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/ct/ct_x509v3.c",
             "crypto/ctype.c",
             "crypto/cversion.c",
+            "crypto/defaults.c",
             "crypto/der_writer.c",
             "crypto/des/cbc_cksm.c",
             "crypto/des/cbc_enc.c",
@@ -436,12 +443,11 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/ec/eck_prn.c",
             "crypto/ec/ecp_mont.c",
             "crypto/ec/ecp_nist.c",
-            // "crypto/ec/ecp_nistp224.c",
-            // "crypto/ec/ecp_nistp256.c",
-            // "crypto/ec/ecp_nistp521.c",
+            "crypto/ec/ecp_nistp224.c",
+            "crypto/ec/ecp_nistp256.c",
+            "crypto/ec/ecp_nistp384.c",
+            "crypto/ec/ecp_nistp521.c",
             "crypto/ec/ecp_nistputil.c",
-            // "crypto/ec/ecp_nistz256.c",
-            // "crypto/ec/ecp_nistz256_table.c",
             "crypto/ec/ecp_oct.c",
             "crypto/ec/ecp_ppc.c",
             "crypto/ec/ecp_smpl.c",
@@ -484,6 +490,7 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/err/err_blocks.c",
             "crypto/err/err_mark.c",
             "crypto/err/err_prn.c",
+            "crypto/err/err_save.c",
             "crypto/ess/ess_asn1.c",
             "crypto/ess/ess_err.c",
             "crypto/ess/ess_lib.c",
@@ -528,7 +535,7 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/evp/evp_err.c",
             "crypto/evp/evp_fetch.c",
             "crypto/evp/evp_key.c",
-            // "crypto/evp/evp_lib.c",
+            "crypto/evp/evp_lib.c",
             "crypto/evp/evp_pbe.c",
             "crypto/evp/evp_pkey.c",
             "crypto/evp/evp_rand.c",
@@ -567,7 +574,9 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/evp/pmeth_check.c",
             "crypto/evp/pmeth_gn.c",
             "crypto/evp/pmeth_lib.c",
-            // "crypto/evp/signature.c",
+            "crypto/evp/signature.c",
+            "crypto/evp/s_lib.c",
+            "crypto/evp/skeymgmt_meth.c",
             "crypto/ex_data.c",
             "crypto/ffc/ffc_backend.c",
             "crypto/ffc/ffc_dh.c",
@@ -578,6 +587,8 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/ffc/ffc_params_validate.c",
             "crypto/getenv.c",
             "crypto/hmac/hmac.c",
+            "crypto/hashtable/hashtable.c",
+            "crypto/hashtable/hashfunc.c",
             "crypto/hpke/hpke.c",
             "crypto/hpke/hpke_util.c",
             "crypto/http/http_client.c",
@@ -588,6 +599,7 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/idea/i_ecb.c",
             "crypto/idea/i_ofb64.c",
             "crypto/idea/i_skey.c",
+            "crypto/indicator_core.c",
             "crypto/info.c",
             "crypto/init.c",
             "crypto/initthread.c",
@@ -705,6 +717,7 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/rand/rand_lib.c",
             "crypto/rand/rand_meth.c",
             "crypto/rand/rand_pool.c",
+            "crypto/rand/rand_uniform.c",
             "crypto/rand/randfile.c",
             "crypto/rc2/rc2_cbc.c",
             "crypto/rc2/rc2_ecb.c",
@@ -768,10 +781,10 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/sm2/sm2_sign.c",
             "crypto/sm3/legacy_sm3.c",
             "crypto/sm3/sm3.c",
-            "crypto/sm4/sm4.c",
             "crypto/sparse_array.c",
             "crypto/srp/srp_lib.c",
             "crypto/srp/srp_vfy.c",
+            "crypto/ssl_err.c",
             "crypto/stack/stack.c",
             "crypto/store/store_err.c",
             "crypto/store/store_init.c",
@@ -787,7 +800,7 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
                 "crypto/thread/arch/thread_win.c"
             else
                 "crypto/thread/arch/thread_posix.c",
-            // "crypto/thread/internal.c",
+            "crypto/thread/internal.c",
             "crypto/threads_lib.c",
             "crypto/threads_none.c",
             if (lib.rootModuleTarget().os.tag == .windows)
@@ -883,6 +896,7 @@ fn libcrypto(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.buil
             "crypto/x509/x509rset.c",
             "crypto/x509/x509spki.c",
             "crypto/x509/x509type.c",
+            "crypto/x509/x509_acert.c",
             "crypto/x509/x_all.c",
             "crypto/x509/x_attrib.c",
             "crypto/x509/x_crl.c",
@@ -990,7 +1004,7 @@ fn libssl(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.builtin
             "ssl/ssl_conf.c",
             "ssl/ssl_err_legacy.c",
             "ssl/ssl_init.c",
-            // "ssl/ssl_lib.c",
+            "ssl/ssl_lib.c",
             "ssl/ssl_mcnf.c",
             "ssl/ssl_rsa.c",
             // "ssl/ssl_rsa_legacy.c",
@@ -1049,6 +1063,10 @@ fn libprovider(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.bu
     lib.root_module.addCMacro("OPENSSL_NO_KTLS", "");
     lib.root_module.addCMacro("OPENSSL_NO_QUIC", "");
     lib.root_module.addCMacro("OPENSSL_CPUID_OBJ", "");
+    lib.root_module.addCMacro("OPENSSL_NO_SM4", "");
+    lib.root_module.addCMacro("OPENSSL_NO_ML_DSA", "");
+    lib.root_module.addCMacro("OPENSSL_NO_SLH_DSA", "");
+    lib.root_module.addCMacro("OPENSSL_NO_ML_KEM", "");
     if (lib.rootModuleTarget().os.tag.isDarwin()) {
         // CommonCrypto
         lib.root_module.linkFramework("CoreServices", .{});
@@ -1108,7 +1126,11 @@ fn libprovider(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.bu
             "providers/implementations/ciphers/cipher_aes.c",
             "providers/implementations/ciphers/cipher_aes_cbc_hmac_sha.c",
             "providers/implementations/ciphers/cipher_aes_cbc_hmac_sha1_hw.c",
+            "providers/implementations/ciphers/cipher_aes_cbc_hmac_sha1_etm_hw.c",
             "providers/implementations/ciphers/cipher_aes_cbc_hmac_sha256_hw.c",
+            "providers/implementations/ciphers/cipher_aes_cbc_hmac_sha256_etm_hw.c",
+            "providers/implementations/ciphers/cipher_aes_cbc_hmac_sha512_etm_hw.c",
+            "providers/implementations/ciphers/cipher_aes_cbc_hmac_sha_etm.c",
             "providers/implementations/ciphers/cipher_aes_ccm.c",
             "providers/implementations/ciphers/cipher_aes_ccm_hw.c",
             "providers/implementations/ciphers/cipher_aes_gcm.c",
@@ -1139,6 +1161,7 @@ fn libprovider(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.bu
             "providers/implementations/ciphers/cipher_cast5_hw.c",
             "providers/implementations/ciphers/cipher_chacha20.c",
             "providers/implementations/ciphers/cipher_chacha20_hw.c",
+            "providers/implementations/ciphers/cipher_chacha20_poly1305.c",
             "providers/implementations/ciphers/cipher_chacha20_poly1305_hw.c",
             "providers/implementations/ciphers/cipher_cts.c",
             "providers/implementations/ciphers/cipher_des.c",
@@ -1158,14 +1181,6 @@ fn libprovider(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.bu
             // "providers/implementations/ciphers/cipher_rc5_hw.c",
             "providers/implementations/ciphers/cipher_seed.c",
             "providers/implementations/ciphers/cipher_seed_hw.c",
-            "providers/implementations/ciphers/cipher_sm4.c",
-            "providers/implementations/ciphers/cipher_sm4_ccm.c",
-            "providers/implementations/ciphers/cipher_sm4_ccm_hw.c",
-            "providers/implementations/ciphers/cipher_sm4_gcm.c",
-            "providers/implementations/ciphers/cipher_sm4_gcm_hw.c",
-            "providers/implementations/ciphers/cipher_sm4_hw.c",
-            "providers/implementations/ciphers/cipher_sm4_xts.c",
-            "providers/implementations/ciphers/cipher_sm4_xts_hw.c",
             "providers/implementations/ciphers/cipher_tdes.c",
             "providers/implementations/ciphers/cipher_tdes_common.c",
             "providers/implementations/ciphers/cipher_tdes_default.c",
@@ -1179,6 +1194,7 @@ fn libprovider(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.bu
             "providers/implementations/ciphers/ciphercommon_gcm.c",
             "providers/implementations/ciphers/ciphercommon_gcm_hw.c",
             "providers/implementations/ciphers/ciphercommon_hw.c",
+            "providers/implementations/ciphers/ciphercommon.c",
             "providers/implementations/digests/blake2_prov.c",
             "providers/implementations/digests/blake2b_prov.c",
             "providers/implementations/digests/blake2s_prov.c",
@@ -1256,7 +1272,9 @@ fn libprovider(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.bu
                 "providers/implementations/rands/seeding/rand_unix.c",
             // "providers/implementations/rands/seeding/rand_vms.c",
             // "providers/implementations/rands/seeding/rand_vxworks.c",
-            // "providers/implementations/rands/test_rng.c",
+            "providers/implementations/rands/test_rng.c",
+            "providers/implementations/skeymgmt/aes_skmgmt.c",
+            "providers/implementations/skeymgmt/generic.c",
             "providers/implementations/signature/dsa_sig.c",
             "providers/implementations/signature/ecdsa_sig.c",
             "providers/implementations/signature/eddsa_sig.c",
diff --git a/crypto/x509/standard_exts.h b/crypto/x509/standard_exts.h
index 63e1db5..e1550fb 100644
--- a/crypto/x509/standard_exts.h
+++ b/crypto/x509/standard_exts.h
@@ -39,7 +39,6 @@ static const X509V3_EXT_METHOD *standard_exts[] = {
 #endif
     &ossl_v3_sxnet,
     &ossl_v3_info,
-    &ossl_v3_audit_identity,
 #ifndef OPENSSL_NO_RFC3779
     &ossl_v3_addr,
     &ossl_v3_asid,
@@ -54,8 +53,6 @@ static const X509V3_EXT_METHOD *standard_exts[] = {
 #endif
     &ossl_v3_sinfo,
     &ossl_v3_policy_constraints,
-    &ossl_v3_targeting_information,
-    &ossl_v3_no_rev_avail,
 #ifndef OPENSSL_NO_OCSP
     &ossl_v3_crl_hold,
 #endif
@@ -63,7 +60,6 @@ static const X509V3_EXT_METHOD *standard_exts[] = {
     &ossl_v3_name_constraints,
     &ossl_v3_policy_mappings,
     &ossl_v3_inhibit_anyp,
-    &ossl_v3_subj_dir_attrs,
     &ossl_v3_idp,
     &ossl_v3_alt[2],
     &ossl_v3_freshest_crl,
@@ -76,26 +72,11 @@ static const X509V3_EXT_METHOD *standard_exts[] = {
     &ossl_v3_issuer_sign_tool,
     &ossl_v3_tls_feature,
     &ossl_v3_ext_admission,
-    &ossl_v3_authority_attribute_identifier,
-    &ossl_v3_role_spec_cert_identifier,
-    &ossl_v3_battcons,
     &ossl_v3_delegated_name_constraints,
-    &ossl_v3_time_specification,
-    &ossl_v3_attribute_descriptor,
-    &ossl_v3_user_notice,
-    &ossl_v3_soa_identifier,
     &ossl_v3_acc_cert_policies,
     &ossl_v3_acc_priv_policies,
-    &ossl_v3_indirect_issuer,
-    &ossl_v3_no_assertion,
     &ossl_v3_aa_issuing_dist_point,
-    &ossl_v3_issued_on_behalf_of,
-    &ossl_v3_single_use,
-    &ossl_v3_group_ac,
-    &ossl_v3_allowed_attribute_assignments,
-    &ossl_v3_attribute_mappings,
     &ossl_v3_holder_name_constraints,
-    &ossl_v3_associated_info,
 };
 
 /* Number of standard extensions */
diff --git a/providers/implementations/ciphers/cipher_chacha20_poly1305.c b/providers/implementations/ciphers/cipher_chacha20_poly1305.c
new file mode 100644
index 0000000..122974a
--- /dev/null
+++ b/providers/implementations/ciphers/cipher_chacha20_poly1305.c
@@ -0,0 +1,475 @@
+/*
+ * Copyright 2019-2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+
+/* Dispatch functions for chacha20_poly1305 cipher */
+
+#include <string.h>
+#include <openssl/proverr.h>
+#include "cipher_chacha20_poly1305.h"
+#include "prov/implementations.h"
+#include "prov/providercommon.h"
+#include "internal/param_names.h"
+
+#define CHACHA20_POLY1305_KEYLEN CHACHA_KEY_SIZE
+#define CHACHA20_POLY1305_BLKLEN 1
+#define CHACHA20_POLY1305_MAX_IVLEN 12
+#define CHACHA20_POLY1305_MODE 0
+#define CHACHA20_POLY1305_FLAGS (PROV_CIPHER_FLAG_AEAD                         \
+                                 | PROV_CIPHER_FLAG_CUSTOM_IV)
+
+static OSSL_FUNC_cipher_newctx_fn chacha20_poly1305_newctx;
+static OSSL_FUNC_cipher_freectx_fn chacha20_poly1305_freectx;
+static OSSL_FUNC_cipher_dupctx_fn chacha20_poly1305_dupctx;
+static OSSL_FUNC_cipher_encrypt_init_fn chacha20_poly1305_einit;
+static OSSL_FUNC_cipher_decrypt_init_fn chacha20_poly1305_dinit;
+static OSSL_FUNC_cipher_get_params_fn chacha20_poly1305_get_params;
+static OSSL_FUNC_cipher_get_ctx_params_fn chacha20_poly1305_get_ctx_params;
+static OSSL_FUNC_cipher_set_ctx_params_fn chacha20_poly1305_set_ctx_params;
+static OSSL_FUNC_cipher_cipher_fn chacha20_poly1305_cipher;
+static OSSL_FUNC_cipher_final_fn chacha20_poly1305_final;
+static OSSL_FUNC_cipher_gettable_ctx_params_fn chacha20_poly1305_gettable_ctx_params;
+static OSSL_FUNC_cipher_settable_ctx_params_fn chacha20_poly1305_settable_ctx_params;
+#define chacha20_poly1305_gettable_params ossl_cipher_generic_gettable_params
+#define chacha20_poly1305_update chacha20_poly1305_cipher
+
+static int chacha20_poly1305_get_ctx_params_find_pidx(const char *);
+
+static void *chacha20_poly1305_newctx(void *provctx)
+{
+    PROV_CHACHA20_POLY1305_CTX *ctx;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx != NULL) {
+        ossl_cipher_generic_initkey(&ctx->base, CHACHA20_POLY1305_KEYLEN * 8,
+                                    CHACHA20_POLY1305_BLKLEN * 8,
+                                    CHACHA20_POLY1305_IVLEN * 8,
+                                    CHACHA20_POLY1305_MODE,
+                                    CHACHA20_POLY1305_FLAGS,
+                                    ossl_prov_cipher_hw_chacha20_poly1305(
+                                        CHACHA20_POLY1305_KEYLEN * 8),
+                                    NULL);
+        ctx->tls_payload_length = NO_TLS_PAYLOAD_LENGTH;
+        ossl_chacha20_initctx(&ctx->chacha);
+    }
+    return ctx;
+}
+
+static void *chacha20_poly1305_dupctx(void *provctx)
+{
+    PROV_CHACHA20_POLY1305_CTX *ctx = provctx;
+    PROV_CHACHA20_POLY1305_CTX *dctx = NULL;
+
+    if (ctx == NULL)
+        return NULL;
+    dctx = OPENSSL_memdup(ctx, sizeof(*ctx));
+    if (dctx != NULL && dctx->base.tlsmac != NULL && dctx->base.alloced) {
+        dctx->base.tlsmac = OPENSSL_memdup(dctx->base.tlsmac,
+                                           dctx->base.tlsmacsize);
+        if (dctx->base.tlsmac == NULL) {
+            OPENSSL_free(dctx);
+            dctx = NULL;
+        }
+    }
+    return dctx;
+}
+
+static void chacha20_poly1305_freectx(void *vctx)
+{
+    PROV_CHACHA20_POLY1305_CTX *ctx = (PROV_CHACHA20_POLY1305_CTX *)vctx;
+
+    if (ctx != NULL) {
+        ossl_cipher_generic_reset_ctx((PROV_CIPHER_CTX *)vctx);
+        OPENSSL_clear_free(ctx, sizeof(*ctx));
+    }
+}
+
+static int chacha20_poly1305_get_params(OSSL_PARAM params[])
+{
+    return ossl_cipher_generic_get_params(params, 0, CHACHA20_POLY1305_FLAGS,
+                                          CHACHA20_POLY1305_KEYLEN * 8,
+                                          CHACHA20_POLY1305_BLKLEN * 8,
+                                          CHACHA20_POLY1305_IVLEN * 8);
+}
+
+static int chacha20_poly1305_get_ctx_params(void *vctx, OSSL_PARAM params[])
+{
+    PROV_CHACHA20_POLY1305_CTX *ctx = (PROV_CHACHA20_POLY1305_CTX *)vctx;
+    OSSL_PARAM *p;
+
+    for (p = params; p->key != NULL; p++)
+        switch (chacha20_poly1305_get_ctx_params_find_pidx(p->key)) {
+        default:
+            break;
+
+        case PIDX_CIPHER_PARAM_IVLEN:
+            if (!OSSL_PARAM_set_size_t(p, CHACHA20_POLY1305_IVLEN)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+                return 0;
+            }
+            break;
+
+        case PIDX_CIPHER_PARAM_KEYLEN:
+            if (!OSSL_PARAM_set_size_t(p, CHACHA20_POLY1305_KEYLEN)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+                return 0;
+            }
+            break;
+
+        case PIDX_CIPHER_PARAM_AEAD_TAGLEN:
+            if (!OSSL_PARAM_set_size_t(p, ctx->tag_len)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+                return 0;
+            }
+            break;
+
+        case PIDX_CIPHER_PARAM_AEAD_TLS1_AAD_PAD:
+            if (!OSSL_PARAM_set_size_t(p, ctx->tls_aad_pad_sz)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+                return 0;
+            }
+            break;
+
+        case PIDX_CIPHER_PARAM_AEAD_TAG:
+            if (p->data_type != OSSL_PARAM_OCTET_STRING) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+                return 0;
+            }
+            if (!ctx->base.enc) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_TAG_NOT_SET);
+                return 0;
+            }
+            if (p->data_size == 0 || p->data_size > POLY1305_BLOCK_SIZE) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_TAG_LENGTH);
+                return 0;
+            }
+            memcpy(p->data, ctx->tag, p->data_size);
+            break;
+        }
+    return 1;
+}
+
+
+/* Machine generated by util/perl/OpenSSL/paramnames.pm */
+static const OSSL_PARAM chacha20_poly1305_known_gettable_ctx_params[] = {
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_KEYLEN, NULL),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_IVLEN, NULL),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_AEAD_TAGLEN, NULL),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TAG, NULL, 0),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_AEAD_TLS1_AAD_PAD, NULL),
+    OSSL_PARAM_END
+};
+
+static int chacha20_poly1305_get_ctx_params_find_pidx(const char *s)
+{
+    switch(s[0]) {
+    default:
+        break;
+    case 'i':
+        if (strcmp("vlen", s + 1) == 0)
+            return PIDX_CIPHER_PARAM_IVLEN;
+        break;
+    case 'k':
+        if (strcmp("eylen", s + 1) == 0)
+            return PIDX_CIPHER_PARAM_KEYLEN;
+        break;
+    case 't':
+        switch(s[1]) {
+        default:
+            break;
+        case 'a':
+            switch(s[2]) {
+            default:
+                break;
+            case 'g':
+                switch(s[3]) {
+                default:
+                    break;
+                case 'l':
+                    if (strcmp("en", s + 4) == 0)
+                        return PIDX_CIPHER_PARAM_AEAD_TAGLEN;
+                    break;
+                case '\0':
+                    return PIDX_CIPHER_PARAM_AEAD_TAG;
+                }
+            }
+            break;
+        case 'l':
+            if (strcmp("saadpad", s + 2) == 0)
+                return PIDX_CIPHER_PARAM_AEAD_TLS1_AAD_PAD;
+        }
+    }
+    return -1;
+}
+/* End of machine generated */
+
+static const OSSL_PARAM *chacha20_poly1305_gettable_ctx_params
+    (ossl_unused void *cctx, ossl_unused void *provctx)
+{
+    return chacha20_poly1305_known_gettable_ctx_params;
+}
+
+/* Machine generated by util/perl/OpenSSL/paramnames.pm */
+static const OSSL_PARAM chacha20_poly1305_known_settable_ctx_params[] = {
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_KEYLEN, NULL),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_IVLEN, NULL),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TAG, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TLS1_AAD, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TLS1_IV_FIXED, NULL, 0),
+    OSSL_PARAM_END
+};
+
+static int chacha20_poly1305_set_ctx_params_find_pidx(const char *s)
+{
+    switch(s[0]) {
+    default:
+        break;
+    case 'i':
+        if (strcmp("vlen", s + 1) == 0)
+            return PIDX_CIPHER_PARAM_IVLEN;
+        break;
+    case 'k':
+        if (strcmp("eylen", s + 1) == 0)
+            return PIDX_CIPHER_PARAM_KEYLEN;
+        break;
+    case 't':
+        switch(s[1]) {
+        default:
+            break;
+        case 'a':
+            if (strcmp("g", s + 2) == 0)
+                return PIDX_CIPHER_PARAM_AEAD_TAG;
+            break;
+        case 'l':
+            switch(s[2]) {
+            default:
+                break;
+            case 's':
+                switch(s[3]) {
+                default:
+                    break;
+                case 'a':
+                    if (strcmp("ad", s + 4) == 0)
+                        return PIDX_CIPHER_PARAM_AEAD_TLS1_AAD;
+                    break;
+                case 'i':
+                    if (strcmp("vfixed", s + 4) == 0)
+                        return PIDX_CIPHER_PARAM_AEAD_TLS1_IV_FIXED;
+                }
+            }
+        }
+    }
+    return -1;
+}
+/* End of machine generated */
+
+static const OSSL_PARAM *chacha20_poly1305_settable_ctx_params(
+        ossl_unused void *cctx, ossl_unused void *provctx
+    )
+{
+    return chacha20_poly1305_known_settable_ctx_params;
+}
+
+static int chacha20_poly1305_set_ctx_params(void *vctx,
+                                            const OSSL_PARAM params[])
+{
+    const OSSL_PARAM *p;
+    size_t len;
+    PROV_CHACHA20_POLY1305_CTX *ctx = (PROV_CHACHA20_POLY1305_CTX *)vctx;
+    PROV_CIPHER_HW_CHACHA20_POLY1305 *hw =
+        (PROV_CIPHER_HW_CHACHA20_POLY1305 *)ctx->base.hw;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    for (p = params; p->key != NULL; p++)
+        switch (chacha20_poly1305_set_ctx_params_find_pidx(p->key)) {
+        default:
+            break;
+
+        case PIDX_CIPHER_PARAM_KEYLEN:
+            if (!OSSL_PARAM_get_size_t(p, &len)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+                return 0;
+            }
+            if (len != CHACHA20_POLY1305_KEYLEN) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+                return 0;
+            }
+            break;
+
+        case PIDX_CIPHER_PARAM_IVLEN:
+            if (!OSSL_PARAM_get_size_t(p, &len)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+                return 0;
+            }
+            if (len != CHACHA20_POLY1305_MAX_IVLEN) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_IV_LENGTH);
+                return 0;
+            }
+            break;
+
+        case PIDX_CIPHER_PARAM_AEAD_TAG:
+            if (p->data_type != OSSL_PARAM_OCTET_STRING) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+                return 0;
+            }
+            if (p->data_size == 0 || p->data_size > POLY1305_BLOCK_SIZE) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_TAG_LENGTH);
+                return 0;
+            }
+            if (p->data != NULL) {
+                if (ctx->base.enc) {
+                    ERR_raise(ERR_LIB_PROV, PROV_R_TAG_NOT_NEEDED);
+                    return 0;
+                }
+                memcpy(ctx->tag, p->data, p->data_size);
+            }
+            ctx->tag_len = p->data_size;
+            break;
+
+        case PIDX_CIPHER_PARAM_AEAD_TLS1_AAD:
+            if (p->data_type != OSSL_PARAM_OCTET_STRING) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+                return 0;
+            }
+            len = hw->tls_init(&ctx->base, p->data, p->data_size);
+            if (len == 0) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DATA);
+                return 0;
+            }
+            ctx->tls_aad_pad_sz = len;
+            break;
+
+        case PIDX_CIPHER_PARAM_AEAD_TLS1_IV_FIXED:
+            if (p->data_type != OSSL_PARAM_OCTET_STRING) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+                return 0;
+            }
+            if (hw->tls_iv_set_fixed(&ctx->base, p->data, p->data_size) == 0) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_IV_LENGTH);
+                return 0;
+            }
+            break;
+        }
+    return 1;
+}
+
+static int chacha20_poly1305_einit(void *vctx, const unsigned char *key,
+                                  size_t keylen, const unsigned char *iv,
+                                  size_t ivlen, const OSSL_PARAM params[])
+{
+    int ret;
+
+    /* The generic function checks for ossl_prov_is_running() */
+    ret = ossl_cipher_generic_einit(vctx, key, keylen, iv, ivlen, NULL);
+    if (ret && iv != NULL) {
+        PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+        PROV_CIPHER_HW_CHACHA20_POLY1305 *hw =
+            (PROV_CIPHER_HW_CHACHA20_POLY1305 *)ctx->hw;
+
+        hw->initiv(ctx);
+    }
+    if (ret && !chacha20_poly1305_set_ctx_params(vctx, params))
+        ret = 0;
+    return ret;
+}
+
+static int chacha20_poly1305_dinit(void *vctx, const unsigned char *key,
+                                  size_t keylen, const unsigned char *iv,
+                                  size_t ivlen, const OSSL_PARAM params[])
+{
+    int ret;
+
+    /* The generic function checks for ossl_prov_is_running() */
+    ret = ossl_cipher_generic_dinit(vctx, key, keylen, iv, ivlen, NULL);
+    if (ret && iv != NULL) {
+        PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+        PROV_CIPHER_HW_CHACHA20_POLY1305 *hw =
+            (PROV_CIPHER_HW_CHACHA20_POLY1305 *)ctx->hw;
+
+        hw->initiv(ctx);
+    }
+    if (ret && !chacha20_poly1305_set_ctx_params(vctx, params))
+        ret = 0;
+    return ret;
+}
+
+static int chacha20_poly1305_cipher(void *vctx, unsigned char *out,
+                                    size_t *outl, size_t outsize,
+                                    const unsigned char *in, size_t inl)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    PROV_CIPHER_HW_CHACHA20_POLY1305 *hw =
+        (PROV_CIPHER_HW_CHACHA20_POLY1305 *)ctx->hw;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (inl == 0) {
+        *outl = 0;
+        return 1;
+    }
+
+    if (outsize < inl) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    if (!hw->aead_cipher(ctx, out, outl, in, inl))
+        return 0;
+
+    return 1;
+}
+
+static int chacha20_poly1305_final(void *vctx, unsigned char *out, size_t *outl,
+                                   size_t outsize)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    PROV_CIPHER_HW_CHACHA20_POLY1305 *hw =
+        (PROV_CIPHER_HW_CHACHA20_POLY1305 *)ctx->hw;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (hw->aead_cipher(ctx, out, outl, NULL, 0) <= 0)
+        return 0;
+
+    *outl = 0;
+    return 1;
+}
+
+/* ossl_chacha20_ossl_poly1305_functions */
+const OSSL_DISPATCH ossl_chacha20_ossl_poly1305_functions[] = {
+    { OSSL_FUNC_CIPHER_NEWCTX, (void (*)(void))chacha20_poly1305_newctx },
+    { OSSL_FUNC_CIPHER_FREECTX, (void (*)(void))chacha20_poly1305_freectx },
+    { OSSL_FUNC_CIPHER_DUPCTX, (void (*)(void))chacha20_poly1305_dupctx },
+    { OSSL_FUNC_CIPHER_ENCRYPT_INIT, (void (*)(void))chacha20_poly1305_einit },
+    { OSSL_FUNC_CIPHER_DECRYPT_INIT, (void (*)(void))chacha20_poly1305_dinit },
+    { OSSL_FUNC_CIPHER_UPDATE, (void (*)(void))chacha20_poly1305_update },
+    { OSSL_FUNC_CIPHER_FINAL, (void (*)(void))chacha20_poly1305_final },
+    { OSSL_FUNC_CIPHER_CIPHER, (void (*)(void))chacha20_poly1305_cipher },
+    { OSSL_FUNC_CIPHER_GET_PARAMS,
+        (void (*)(void))chacha20_poly1305_get_params },
+    { OSSL_FUNC_CIPHER_GETTABLE_PARAMS,
+        (void (*)(void))chacha20_poly1305_gettable_params },
+    { OSSL_FUNC_CIPHER_GET_CTX_PARAMS,
+         (void (*)(void))chacha20_poly1305_get_ctx_params },
+    { OSSL_FUNC_CIPHER_GETTABLE_CTX_PARAMS,
+        (void (*)(void))chacha20_poly1305_gettable_ctx_params },
+    { OSSL_FUNC_CIPHER_SET_CTX_PARAMS,
+        (void (*)(void))chacha20_poly1305_set_ctx_params },
+    { OSSL_FUNC_CIPHER_SETTABLE_CTX_PARAMS,
+        (void (*)(void))chacha20_poly1305_settable_ctx_params },
+    OSSL_DISPATCH_END
+};
+
diff --git a/providers/implementations/ciphers/ciphercommon.c b/providers/implementations/ciphers/ciphercommon.c
new file mode 100644
index 0000000..c07b447
--- /dev/null
+++ b/providers/implementations/ciphers/ciphercommon.c
@@ -0,0 +1,881 @@
+/*
+ * Copyright 2019-2025 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+
+/*
+ * Generic dispatch table functions for ciphers.
+ */
+
+/* For SSL3_VERSION */
+#include <string.h>
+#include <openssl/prov_ssl.h>
+#include <openssl/proverr.h>
+#include "ciphercommon_local.h"
+#include "prov/provider_ctx.h"
+#include "prov/providercommon.h"
+#include "internal/skey.h"
+#include "internal/e_os.h"
+#include "internal/param_names.h"
+#include "crypto/types.h"
+
+/*-
+ * Generic cipher functions for OSSL_PARAM gettables and settables
+ */
+static const OSSL_PARAM cipher_known_gettable_params[] = {
+    OSSL_PARAM_uint(OSSL_CIPHER_PARAM_MODE, NULL),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_KEYLEN, NULL),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_IVLEN, NULL),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_BLOCK_SIZE, NULL),
+    OSSL_PARAM_int(OSSL_CIPHER_PARAM_AEAD, NULL),
+    OSSL_PARAM_int(OSSL_CIPHER_PARAM_CUSTOM_IV, NULL),
+    OSSL_PARAM_int(OSSL_CIPHER_PARAM_CTS, NULL),
+    OSSL_PARAM_int(OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK, NULL),
+    OSSL_PARAM_int(OSSL_CIPHER_PARAM_HAS_RAND_KEY, NULL),
+    OSSL_PARAM_END
+};
+const OSSL_PARAM *ossl_cipher_generic_gettable_params(ossl_unused void *provctx)
+{
+    return cipher_known_gettable_params;
+}
+
+int ossl_cipher_generic_get_params(OSSL_PARAM params[], unsigned int md,
+                                   uint64_t flags,
+                                   size_t kbits, size_t blkbits, size_t ivbits)
+{
+    OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_MODE);
+    if (p != NULL && !OSSL_PARAM_set_uint(p, md)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_AEAD);
+    if (p != NULL
+        && !OSSL_PARAM_set_int(p, (flags & PROV_CIPHER_FLAG_AEAD) != 0)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_CUSTOM_IV);
+    if (p != NULL
+        && !OSSL_PARAM_set_int(p, (flags & PROV_CIPHER_FLAG_CUSTOM_IV) != 0)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_CTS);
+    if (p != NULL
+        && !OSSL_PARAM_set_int(p, (flags & PROV_CIPHER_FLAG_CTS) != 0)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK);
+    if (p != NULL
+        && !OSSL_PARAM_set_int(p, (flags & PROV_CIPHER_FLAG_TLS1_MULTIBLOCK) != 0)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_HAS_RAND_KEY);
+    if (p != NULL
+        && !OSSL_PARAM_set_int(p, (flags & PROV_CIPHER_FLAG_RAND_KEY) != 0)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_ENCRYPT_THEN_MAC);
+    if (p != NULL
+        && !OSSL_PARAM_set_int(p, (flags & EVP_CIPH_FLAG_ENC_THEN_MAC) != 0)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_KEYLEN);
+    if (p != NULL && !OSSL_PARAM_set_size_t(p, kbits / 8)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_BLOCK_SIZE);
+    if (p != NULL && !OSSL_PARAM_set_size_t(p, blkbits / 8)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_IVLEN);
+    if (p != NULL && !OSSL_PARAM_set_size_t(p, ivbits / 8)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    return 1;
+}
+
+CIPHER_DEFAULT_GETTABLE_CTX_PARAMS_START(ossl_cipher_generic)
+{ OSSL_CIPHER_PARAM_TLS_MAC, OSSL_PARAM_OCTET_PTR, NULL, 0, OSSL_PARAM_UNMODIFIED },
+CIPHER_DEFAULT_GETTABLE_CTX_PARAMS_END(ossl_cipher_generic)
+
+CIPHER_DEFAULT_SETTABLE_CTX_PARAMS_START(ossl_cipher_generic)
+OSSL_PARAM_uint(OSSL_CIPHER_PARAM_USE_BITS, NULL),
+OSSL_PARAM_uint(OSSL_CIPHER_PARAM_TLS_VERSION, NULL),
+OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_TLS_MAC_SIZE, NULL),
+CIPHER_DEFAULT_SETTABLE_CTX_PARAMS_END(ossl_cipher_generic)
+
+/*
+ * Variable key length cipher functions for OSSL_PARAM settables
+ */
+int ossl_cipher_var_keylen_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    const OSSL_PARAM *p;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    if (!ossl_cipher_generic_set_ctx_params(vctx, params))
+        return 0;
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_KEYLEN);
+    if (p != NULL) {
+        size_t keylen;
+
+        if (!OSSL_PARAM_get_size_t(p, &keylen)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        if (ctx->keylen != keylen) {
+            ctx->keylen = keylen;
+            ctx->key_set = 0;
+        }
+    }
+    return 1;
+}
+
+CIPHER_DEFAULT_SETTABLE_CTX_PARAMS_START(ossl_cipher_var_keylen)
+OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_KEYLEN, NULL),
+CIPHER_DEFAULT_SETTABLE_CTX_PARAMS_END(ossl_cipher_var_keylen)
+
+/*-
+ * AEAD cipher functions for OSSL_PARAM gettables and settables
+ */
+
+/* Machine generated by util/perl/OpenSSL/paramnames.pm */
+static const OSSL_PARAM cipher_aead_known_gettable_ctx_params[] = {
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_KEYLEN, NULL),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_IVLEN, NULL),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_AEAD_TAGLEN, NULL),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_IV, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_UPDATED_IV, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TAG, NULL, 0),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_AEAD_TLS1_AAD_PAD, NULL),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TLS1_GET_IV_GEN, NULL, 0),
+    OSSL_PARAM_uint(OSSL_CIPHER_PARAM_AEAD_IV_GENERATED, NULL),
+    OSSL_PARAM_END
+};
+
+int ossl_cipher_aead_get_ctx_params_find_pidx(const char *s)
+{
+    switch(s[0]) {
+    default:
+        break;
+    case 'i':
+        switch(s[1]) {
+        default:
+            break;
+        case 'v':
+            switch(s[2]) {
+            default:
+                break;
+            case '-':
+                if (strcmp("generated", s + 3) == 0)
+                    return PIDX_CIPHER_PARAM_AEAD_IV_GENERATED;
+                break;
+            case 'l':
+                if (strcmp("en", s + 3) == 0)
+                    return PIDX_CIPHER_PARAM_IVLEN;
+                break;
+            case '\0':
+                return PIDX_CIPHER_PARAM_IV;
+            }
+        }
+        break;
+    case 'k':
+        if (strcmp("eylen", s + 1) == 0)
+            return PIDX_CIPHER_PARAM_KEYLEN;
+        break;
+    case 't':
+        switch(s[1]) {
+        default:
+            break;
+        case 'a':
+            switch(s[2]) {
+            default:
+                break;
+            case 'g':
+                switch(s[3]) {
+                default:
+                    break;
+                case 'l':
+                    if (strcmp("en", s + 4) == 0)
+                        return PIDX_CIPHER_PARAM_AEAD_TAGLEN;
+                    break;
+                case '\0':
+                    return PIDX_CIPHER_PARAM_AEAD_TAG;
+                }
+            }
+            break;
+        case 'l':
+            switch(s[2]) {
+            default:
+                break;
+            case 's':
+                switch(s[3]) {
+                default:
+                    break;
+                case 'a':
+                    if (strcmp("adpad", s + 4) == 0)
+                        return PIDX_CIPHER_PARAM_AEAD_TLS1_AAD_PAD;
+                    break;
+                case 'i':
+                    if (strcmp("vgen", s + 4) == 0)
+                        return PIDX_CIPHER_PARAM_AEAD_TLS1_GET_IV_GEN;
+                }
+            }
+        }
+        break;
+    case 'u':
+        if (strcmp("pdated-iv", s + 1) == 0)
+            return PIDX_CIPHER_PARAM_UPDATED_IV;
+    }
+    return -1;
+}
+/* End of machine generated */
+
+const OSSL_PARAM *ossl_cipher_aead_gettable_ctx_params(
+        ossl_unused void *cctx, ossl_unused void *provctx
+    )
+{
+    return cipher_aead_known_gettable_ctx_params;
+}
+
+/* Machine generated by util/perl/OpenSSL/paramnames.pm */
+static const OSSL_PARAM cipher_aead_known_settable_ctx_params[] = {
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_AEAD_IVLEN, NULL),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TAG, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TLS1_AAD, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TLS1_IV_FIXED, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_AEAD_TLS1_SET_IV_INV, NULL, 0),
+    OSSL_PARAM_END
+};
+
+int ossl_cipher_aead_set_ctx_params_find_pidx(const char *s)
+{
+    switch(s[0]) {
+    default:
+        break;
+    case 'i':
+        if (strcmp("vlen", s + 1) == 0)
+            return PIDX_CIPHER_PARAM_AEAD_IVLEN;
+        break;
+    case 't':
+        switch(s[1]) {
+        default:
+            break;
+        case 'a':
+            if (strcmp("g", s + 2) == 0)
+                return PIDX_CIPHER_PARAM_AEAD_TAG;
+            break;
+        case 'l':
+            switch(s[2]) {
+            default:
+                break;
+            case 's':
+                switch(s[3]) {
+                default:
+                    break;
+                case 'a':
+                    if (strcmp("ad", s + 4) == 0)
+                        return PIDX_CIPHER_PARAM_AEAD_TLS1_AAD;
+                    break;
+                case 'i':
+                    switch(s[4]) {
+                    default:
+                        break;
+                    case 'v':
+                        switch(s[5]) {
+                        default:
+                            break;
+                        case 'f':
+                            if (strcmp("ixed", s + 6) == 0)
+                                return PIDX_CIPHER_PARAM_AEAD_TLS1_IV_FIXED;
+                            break;
+                        case 'i':
+                            if (strcmp("nv", s + 6) == 0)
+                                return PIDX_CIPHER_PARAM_AEAD_TLS1_SET_IV_INV;
+                        }
+                    }
+                }
+            }
+        }
+    }
+    return -1;
+}
+/* End of machine generated */
+
+const OSSL_PARAM *ossl_cipher_aead_settable_ctx_params(
+        ossl_unused void *cctx, ossl_unused void *provctx
+    )
+{
+    return cipher_aead_known_settable_ctx_params;
+}
+
+void ossl_cipher_generic_reset_ctx(PROV_CIPHER_CTX *ctx)
+{
+    if (ctx != NULL && ctx->alloced) {
+        OPENSSL_free(ctx->tlsmac);
+        ctx->alloced = 0;
+        ctx->tlsmac = NULL;
+    }
+}
+
+static int cipher_generic_init_internal(PROV_CIPHER_CTX *ctx,
+                                        const unsigned char *key, size_t keylen,
+                                        const unsigned char *iv, size_t ivlen,
+                                        const OSSL_PARAM params[], int enc)
+{
+    ctx->num = 0;
+    ctx->bufsz = 0;
+    ctx->updated = 0;
+    ctx->enc = enc ? 1 : 0;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (iv != NULL && ctx->mode != EVP_CIPH_ECB_MODE) {
+        if (!ossl_cipher_generic_initiv(ctx, iv, ivlen))
+            return 0;
+    }
+    if (iv == NULL && ctx->iv_set
+        && (ctx->mode == EVP_CIPH_CBC_MODE
+            || ctx->mode == EVP_CIPH_CFB_MODE
+            || ctx->mode == EVP_CIPH_OFB_MODE))
+        /* reset IV for these modes to keep compatibility with 1.1.1 */
+        memcpy(ctx->iv, ctx->oiv, ctx->ivlen);
+
+    if (key != NULL) {
+        if (ctx->variable_keylength == 0) {
+            if (keylen != ctx->keylen) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+                return 0;
+            }
+        } else {
+            ctx->keylen = keylen;
+        }
+        if (!ctx->hw->init(ctx, key, ctx->keylen))
+            return 0;
+        ctx->key_set = 1;
+    }
+    return ossl_cipher_generic_set_ctx_params(ctx, params);
+}
+
+int ossl_cipher_generic_einit(void *vctx, const unsigned char *key,
+                              size_t keylen, const unsigned char *iv,
+                              size_t ivlen, const OSSL_PARAM params[])
+{
+    return cipher_generic_init_internal((PROV_CIPHER_CTX *)vctx, key, keylen,
+                                        iv, ivlen, params, 1);
+}
+
+int ossl_cipher_generic_dinit(void *vctx, const unsigned char *key,
+                              size_t keylen, const unsigned char *iv,
+                              size_t ivlen, const OSSL_PARAM params[])
+{
+    return cipher_generic_init_internal((PROV_CIPHER_CTX *)vctx, key, keylen,
+                                        iv, ivlen, params, 0);
+}
+
+int ossl_cipher_generic_skey_einit(void *vctx, void *skeydata,
+                                   const unsigned char *iv, size_t ivlen,
+                                   const OSSL_PARAM params[])
+{
+    PROV_SKEY *key = skeydata;
+
+    return cipher_generic_init_internal((PROV_CIPHER_CTX *)vctx,
+                                        key->data, key->length,
+                                        iv, ivlen, params, 1);
+}
+
+int ossl_cipher_generic_skey_dinit(void *vctx, void *skeydata,
+                                   const unsigned char *iv, size_t ivlen,
+                                   const OSSL_PARAM params[])
+{
+    PROV_SKEY *key = skeydata;
+
+    return cipher_generic_init_internal((PROV_CIPHER_CTX *)vctx,
+                                        key->data, key->length,
+                                        iv, ivlen, params, 0);
+}
+
+/* Max padding including padding length byte */
+#define MAX_PADDING 256
+
+int ossl_cipher_generic_block_update(void *vctx, unsigned char *out,
+                                     size_t *outl, size_t outsize,
+                                     const unsigned char *in, size_t inl)
+{
+    size_t outlint = 0;
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    size_t blksz = ctx->blocksize;
+    size_t nextblocks;
+
+    if (!ctx->key_set) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+        return 0;
+    }
+
+    if (ctx->tlsversion > 0) {
+        /*
+         * Each update call corresponds to a TLS record and is individually
+         * padded
+         */
+
+        /* Sanity check inputs */
+        if (in == NULL
+                || in != out
+                || outsize < inl
+                || !ctx->pad) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+            return 0;
+        }
+
+        if (ctx->enc) {
+            unsigned char padval;
+            size_t padnum, loop;
+
+            /* Add padding */
+
+            padnum = blksz - (inl % blksz);
+
+            if (outsize < inl + padnum) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+                return 0;
+            }
+
+            if (padnum > MAX_PADDING) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+                return 0;
+            }
+            padval = (unsigned char)(padnum - 1);
+            if (ctx->tlsversion == SSL3_VERSION) {
+                if (padnum > 1)
+                    memset(out + inl, 0, padnum - 1);
+                *(out + inl + padnum - 1) = padval;
+            } else {
+                /* we need to add 'padnum' padding bytes of value padval */
+                for (loop = inl; loop < inl + padnum; loop++)
+                    out[loop] = padval;
+            }
+            inl += padnum;
+        }
+
+        if ((inl % blksz) != 0) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+            return 0;
+        }
+
+
+        /* Shouldn't normally fail */
+        if (!ctx->hw->cipher(ctx, out, in, inl)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+            return 0;
+        }
+
+        if (ctx->alloced) {
+            OPENSSL_free(ctx->tlsmac);
+            ctx->alloced = 0;
+            ctx->tlsmac = NULL;
+        }
+
+        /* This only fails if padding is publicly invalid */
+        *outl = inl;
+        if (!ctx->enc
+            && !ossl_cipher_tlsunpadblock(ctx->libctx, ctx->tlsversion,
+                                          out, outl,
+                                          blksz, &ctx->tlsmac, &ctx->alloced,
+                                          ctx->tlsmacsize, 0)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+            return 0;
+        }
+        return 1;
+    }
+
+    if (ctx->bufsz != 0)
+        nextblocks = ossl_cipher_fillblock(ctx->buf, &ctx->bufsz, blksz,
+                                           &in, &inl);
+    else
+        nextblocks = inl & ~(blksz-1);
+
+    /*
+     * If we're decrypting and we end an update on a block boundary we hold
+     * the last block back in case this is the last update call and the last
+     * block is padded.
+     */
+    if (ctx->bufsz == blksz && (ctx->enc || inl > 0 || !ctx->pad)) {
+        if (outsize < blksz) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+            return 0;
+        }
+        if (!ctx->hw->cipher(ctx, out, ctx->buf, blksz)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+            return 0;
+        }
+        ctx->bufsz = 0;
+        outlint = blksz;
+        out += blksz;
+    }
+    if (nextblocks > 0) {
+        if (!ctx->enc && ctx->pad && nextblocks == inl) {
+            if (!ossl_assert(inl >= blksz)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+                return 0;
+            }
+            nextblocks -= blksz;
+        }
+        outlint += nextblocks;
+        if (outsize < outlint) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+            return 0;
+        }
+    }
+    if (nextblocks > 0) {
+        if (!ctx->hw->cipher(ctx, out, in, nextblocks)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+            return 0;
+        }
+        in += nextblocks;
+        inl -= nextblocks;
+    }
+    if (inl != 0
+        && !ossl_cipher_trailingdata(ctx->buf, &ctx->bufsz, blksz, &in, &inl)) {
+        /* ERR_raise already called */
+        return 0;
+    }
+
+    *outl = outlint;
+    return inl == 0;
+}
+
+int ossl_cipher_generic_block_final(void *vctx, unsigned char *out,
+                                    size_t *outl, size_t outsize)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    size_t blksz = ctx->blocksize;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (!ctx->key_set) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+        return 0;
+    }
+
+    if (ctx->tlsversion > 0) {
+        /* We never finalize TLS, so this is an error */
+        ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+        return 0;
+    }
+
+    if (ctx->enc) {
+        if (ctx->pad) {
+            ossl_cipher_padblock(ctx->buf, &ctx->bufsz, blksz);
+        } else if (ctx->bufsz == 0) {
+            *outl = 0;
+            return 1;
+        } else if (ctx->bufsz != blksz) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_WRONG_FINAL_BLOCK_LENGTH);
+            return 0;
+        }
+
+        if (outsize < blksz) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+            return 0;
+        }
+        if (!ctx->hw->cipher(ctx, out, ctx->buf, blksz)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+            return 0;
+        }
+        ctx->bufsz = 0;
+        *outl = blksz;
+        return 1;
+    }
+
+    /* Decrypting */
+    if (ctx->bufsz != blksz) {
+        if (ctx->bufsz == 0 && !ctx->pad) {
+            *outl = 0;
+            return 1;
+        }
+        ERR_raise(ERR_LIB_PROV, PROV_R_WRONG_FINAL_BLOCK_LENGTH);
+        return 0;
+    }
+
+    if (!ctx->hw->cipher(ctx, ctx->buf, ctx->buf, blksz)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+        return 0;
+    }
+
+    if (ctx->pad && !ossl_cipher_unpadblock(ctx->buf, &ctx->bufsz, blksz)) {
+        /* ERR_raise already called */
+        return 0;
+    }
+
+    if (outsize < ctx->bufsz) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return 0;
+    }
+    memcpy(out, ctx->buf, ctx->bufsz);
+    *outl = ctx->bufsz;
+    ctx->bufsz = 0;
+    return 1;
+}
+
+int ossl_cipher_generic_stream_update(void *vctx, unsigned char *out,
+                                      size_t *outl, size_t outsize,
+                                      const unsigned char *in, size_t inl)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+
+    if (!ctx->key_set) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+        return 0;
+    }
+
+    if (inl == 0) {
+        *outl = 0;
+        return 1;
+    }
+
+    if (outsize < inl) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    if (!ctx->hw->cipher(ctx, out, in, inl)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+        return 0;
+    }
+
+    *outl = inl;
+    if (!ctx->enc && ctx->tlsversion > 0) {
+        /*
+        * Remove any TLS padding. Only used by cipher_aes_cbc_hmac_sha1_hw.c and
+        * cipher_aes_cbc_hmac_sha256_hw.c
+        */
+        if (ctx->removetlspad) {
+            /*
+             * We should have already failed in the cipher() call above if this
+             * isn't true.
+             */
+            if (!ossl_assert(*outl >= (size_t)(out[inl - 1] + 1)))
+                return 0;
+            /* The actual padding length */
+            *outl -= out[inl - 1] + 1;
+        }
+
+        /* TLS MAC and explicit IV if relevant. We should have already failed
+         * in the cipher() call above if *outl is too short.
+         */
+        if (!ossl_assert(*outl >= ctx->removetlsfixed))
+            return 0;
+        *outl -= ctx->removetlsfixed;
+
+        /* Extract the MAC if there is one */
+        if (ctx->tlsmacsize > 0) {
+            if (*outl < ctx->tlsmacsize)
+                return 0;
+
+            ctx->tlsmac = out + *outl - ctx->tlsmacsize;
+            *outl -= ctx->tlsmacsize;
+        }
+    }
+
+    return 1;
+}
+int ossl_cipher_generic_stream_final(void *vctx, unsigned char *out,
+                                     size_t *outl, size_t outsize)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (!ctx->key_set) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+        return 0;
+    }
+
+    *outl = 0;
+    return 1;
+}
+
+int ossl_cipher_generic_cipher(void *vctx, unsigned char *out, size_t *outl,
+                               size_t outsize, const unsigned char *in,
+                               size_t inl)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (!ctx->key_set) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+        return 0;
+    }
+
+    if (outsize < inl) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    if (!ctx->hw->cipher(ctx, out, in, inl)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+        return 0;
+    }
+
+    *outl = inl;
+    return 1;
+}
+
+int ossl_cipher_generic_get_ctx_params(void *vctx, OSSL_PARAM params[])
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_IVLEN);
+    if (p != NULL && !OSSL_PARAM_set_size_t(p, ctx->ivlen)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_PADDING);
+    if (p != NULL && !OSSL_PARAM_set_uint(p, ctx->pad)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_IV);
+    if (p != NULL
+        && !OSSL_PARAM_set_octet_string_or_ptr(p, ctx->oiv, ctx->ivlen)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_UPDATED_IV);
+    if (p != NULL
+        && !OSSL_PARAM_set_octet_string_or_ptr(p, ctx->iv, ctx->ivlen)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_NUM);
+    if (p != NULL && !OSSL_PARAM_set_uint(p, ctx->num)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_KEYLEN);
+    if (p != NULL && !OSSL_PARAM_set_size_t(p, ctx->keylen)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_TLS_MAC);
+    if (p != NULL
+        && !OSSL_PARAM_set_octet_ptr(p, ctx->tlsmac, ctx->tlsmacsize)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    return 1;
+}
+
+int ossl_cipher_generic_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    const OSSL_PARAM *p;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_PADDING);
+    if (p != NULL) {
+        unsigned int pad;
+
+        if (!OSSL_PARAM_get_uint(p, &pad)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        ctx->pad = pad ? 1 : 0;
+    }
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_USE_BITS);
+    if (p != NULL) {
+        unsigned int bits;
+
+        if (!OSSL_PARAM_get_uint(p, &bits)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        ctx->use_bits = bits ? 1 : 0;
+    }
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_TLS_VERSION);
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_uint(p, &ctx->tlsversion)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+    }
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_TLS_MAC_SIZE);
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_size_t(p, &ctx->tlsmacsize)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+    }
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_NUM);
+    if (p != NULL) {
+        unsigned int num;
+
+        if (!OSSL_PARAM_get_uint(p, &num)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        ctx->num = num;
+    }
+    return 1;
+}
+
+int ossl_cipher_generic_initiv(PROV_CIPHER_CTX *ctx, const unsigned char *iv,
+                               size_t ivlen)
+{
+    if (ivlen != ctx->ivlen
+        || ivlen > sizeof(ctx->iv)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_IV_LENGTH);
+        return 0;
+    }
+    ctx->iv_set = 1;
+    memcpy(ctx->iv, iv, ivlen);
+    memcpy(ctx->oiv, iv, ivlen);
+    return 1;
+}
+
+void ossl_cipher_generic_initkey(void *vctx, size_t kbits, size_t blkbits,
+                                 size_t ivbits, unsigned int mode,
+                                 uint64_t flags, const PROV_CIPHER_HW *hw,
+                                 void *provctx)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+
+    if ((flags & PROV_CIPHER_FLAG_INVERSE_CIPHER) != 0)
+        ctx->inverse_cipher = 1;
+    if ((flags & PROV_CIPHER_FLAG_VARIABLE_LENGTH) != 0)
+        ctx->variable_keylength = 1;
+
+    ctx->pad = 1;
+    ctx->keylen = ((kbits) / 8);
+    ctx->ivlen = ((ivbits) / 8);
+    ctx->hw = hw;
+    ctx->mode = mode;
+    ctx->blocksize = blkbits / 8;
+    if (provctx != NULL)
+        ctx->libctx = PROV_LIBCTX_OF(provctx); /* used for rand */
+}
diff --git a/ssl/s3_lib.c b/ssl/s3_lib.c
index 94e499c..bebd159 100644
--- a/ssl/s3_lib.c
+++ b/ssl/s3_lib.c
@@ -3442,7 +3442,9 @@ void ssl3_free(SSL *s)
     ssl3_free_digest_list(sc);
     OPENSSL_free(sc->s3.alpn_selected);
     OPENSSL_free(sc->s3.alpn_proposed);
+#ifndef OPENSSL_NO_QUIC
     ossl_quic_tls_free(sc->qtls);
+#endif
 
 #ifndef OPENSSL_NO_PSK
     OPENSSL_free(sc->s3.tmp.psk);
diff --git a/ssl/ssl_lib.c b/ssl/ssl_lib.c
index a1a56b9..f36511b 100644
--- a/ssl/ssl_lib.c
+++ b/ssl/ssl_lib.c
@@ -652,7 +652,9 @@ int ossl_ssl_connection_reset(SSL *s)
             return 0;
     }
 
+    #ifndef OPENSSL_NO_QUIC
     ossl_quic_tls_clear(sc->qtls);
+#endif
 
     if (!RECORD_LAYER_reset(&sc->rlayer))
         return 0;
